src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -11,80 +11,507 @@
+    ucc_rank_t         size        = coll_task->team->params.size;
+    ucc_datatype_t     dtype       = TASK_ARGS(task).dst.info.datatype;
+    size_t             dt_size     = ucc_dt_size(dtype);
+    int                put_window_size = UCC_TL_UCP_TEAM_LIB(team)
+                                 ->cfg.allreduce_sliding_window_put_window_size;
+    ucc_tl_ucp_allreduce_sw_pipeline_t *pipe =
+        task->allreduce_sliding_window.pipe;
+    ucc_tl_ucp_allreduce_sw_host_allgather_t *allgather_data =
+        task->allreduce_sliding_window.allgather_data;
+    size_t allgather_size = sizeof(ucc_tl_ucp_allreduce_sw_host_allgather_t);
+
+    ucc_base_coll_args_t bargs    = {
+        .args = {
+            .coll_type = UCC_COLL_TYPE_ALLGATHER,
+            .mask      = 0,
+            .src.info = {.buffer   = allgather_data,
+                         .count    = allgather_size,
+                         .datatype = UCC_DT_UINT8,
+                         .mem_type = UCC_MEMORY_TYPE_HOST},
+            .dst.info = {.buffer   = PTR_OFFSET(allgather_data, allgather_size),
+                         .count    = allgather_size * size,
+                         .datatype = UCC_DT_UINT8,
+                         .mem_type = UCC_MEMORY_TYPE_HOST}
+        }
+    };
+
+    UCC_CHECK_GOTO(ucc_tl_ucp_allgather_knomial_init(&bargs,
+                    &team->super.super,
+                    &task->allreduce_sliding_window.allgather_task),
+        out, status);
+
+    task->allreduce_sliding_window.allgather_task->executor =
+                                        task->allreduce_sliding_window.executor;
+
+    UCC_CHECK_GOTO(ucc_tl_ucp_allgather_knomial_start(
+                    task->allreduce_sliding_window.allgather_task),
+        out, status);
+
+    if (put_window_size <= 0)
+        put_window_size = size;
+
+    ucc_tl_ucp_allreduce_sliding_window_reset_pipeline(
+        pipe, rank, put_window_size);
+
+    pipe->my_count  = count_total / size;
+    pipe->my_offset = pipe->my_count * dt_size * rank;
+    if (rank == size - 1) {
+        pipe->my_count += count_total % size;
+    }
+
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+
+    task->allreduce_sliding_window.reduce_in_progress = 0;
+    task->allreduce_sliding_window.barrier_task       = NULL;
+
+    return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+
+out:
+    ucc_tl_ucp_coll_finalize(task->allreduce_sliding_window.allgather_task);
+    tl_error(UCC_TASK_LIB(task), "failed to start allreduce sliding window");
+    return status;
 }
 
 ucc_status_t
-ucc_tl_ucp_allreduce_sliding_window_finalize(ucc_coll_task_t __attribute__((unused)) *coll_task) //NOLINT
+ucc_tl_ucp_allreduce_sliding_window_finalize(ucc_coll_task_t *coll_task)
 {
-    return UCC_OK;
+    ucc_tl_ucp_task_t *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_status_t       st =
+        ucc_ee_executor_finalize(task->allreduce_sliding_window.executor);
```

Comment:
I had created the executor for the allgather task since I found that the allreduce task did not have one. Which function should free it? I don't see how this is freed unless I call `ucc_ee_executor_finalize` myself here.

Reply:
> Which function should free it? I don't see how this is freed unless I call `ucc_ee_executor_finalize` myself here.

I was thinking about this

https://github.com/openucx/ucc/blob/c65c4a3fb1732799b1498ce97b148f0acfe490bd/src/schedule/ucc_schedule.h#L212

> I had created the executor for the allgather task since I found that the allreduce task did not have one

For creation this should be handled here

https://github.com/openucx/ucc/blob/c65c4a3fb1732799b1498ce97b148f0acfe490bd/src/core/ucc_coll.c#L266

provided that `task->flags` has `UCC_COLL_TASK_FLAG_EXECUTOR`

Does it make sense?
@Sergei-Lebedev could you confirm that this is should be the workflow for the executor ?

Reply:
After setting the flag, I confirmed it works by allocating in `ucc_collective_init` and freeing in `ucc_collective_finalize_internal`. Thank you Sam!!